cartographer/mapping/map_builder.cc

```diff
@@ -174,11 +174,12 @@ void MapBuilder::SerializeState(io::ProtoStreamWriter* const writer) {
 }
 
 void MapBuilder::LoadMap(io::ProtoStreamReader* const reader) {
-  proto::SparsePoseGraph pose_graph;
-  CHECK(reader->ReadProto(&pose_graph));
+  const std::shared_ptr<proto::SparsePoseGraph> pose_graph =
```

Comment:
Using a `shared_ptr` to avoid copy of this comparatively small amount of data seems like premature optimization, especially given that we copy the constraints in other parts of the code.

Reply:
According to #433, it can be quite big.

Reply:
Right, but using a `shared_ptr` here does not really help with that problem. If we wanted to properly support the use case of deserializing 100s of MBs of constraints we would need to move the constraints out of the `proto::SparsePoseGraph` and support loading them in chunks. So let's do a copy here.

BTW personally I would only use a `shared_ptr` if the lifetime of the object is known only at runtime. I would always opt for `unique_ptr`.

Reply:
You can't easily move a `unique_ptr` into a lambda in C++11. You would have to release it and capture a regular pointer. And we don't exactly know how long the function is going to sit around on the worker queue (does this fit your definition of knowing it only at runtime)?

But sure, I can change it here to copying the SPG proto when adding the function to the worker queue, if you prefer.

Reply:
Cool. Let's do that and get this thing merged.